### ISSUE TITLE
Handling component focus with providers

### ISSUE THREAD

**Issue by FredyC** [loggur/react-redux-provide#39]
By focus I mean essentially to have a component behave differently if it has a focus. From basic input which should set true DOM focus there, but also eg. list of choices that should be navigated by keyboard only if there is a focus. On top of that each type of focus can come with additional settings (like cursor position within input).

Classical redux approach is pretty obvious, just some enum of all possible identifiers for focusable elements and dispatch action with those when it needs to be set there. I guess it works just fine if there is not much of these, but for some larger amount it might be tiring to maintain.

One solution I can think of is have a provider to dynamically create actions and reducers based on list of identifiers. So then I will have like `focusBigInput` and `hasFocusBigInput`. It's probably better compared to regular solution, but it's still somewhat tighly coupled and provider needs to know about all focusable components.

The best would be if every component can simply define propTypes `hasFocus` and `changeFocus` and it be automatically recognized per a component. Is it possible to achieve something like that?

**Comment by timbur** [comment]
Interesting.  I'm not sure if this is something for providers, but hey, anything is possible!  😅

It seems like this should be something internal to the components themselves, using `setState`, although yeah that wouldn't be possible with stateless functional components, of course.  I think I know what you're asking and I think it is possible, but I would need to see an example of what exactly you mean before I can make a real assessment.

**Comment by FredyC** [comment]
Simple component like this...
```
import ReactStamp from 'core/reactStamp';

export default ReactStamp.compose({
	propTypes: {
		hasFocus: PT.bool.isRequired,
		setFocus: PT.func.isRequired,
	},
	componentDidMount() {
		this.updateFocus();
	}
	componentDidUpdate() {
		this.updateFocus();
	}
	updateFocus() {
		if (this.props.hasFocus) {
			this.findDOMNode().focus();
		}
	}
	render() {
		return <input />;
	},
});
```
With magic coming from stamps I can even have this as a composable behavior if names of those props would be same. 

Actually might it be possible with dynamic `key`? I could define another prop there like `focusId` so for every component I would get different provider instance... Hey I think I am starting to get this more and more :)

**Comment by timbur** [comment]
Nice!  Yes it sounds like a dynamic `key` would be the way to go!  😁

**Comment by FredyC** [comment]
Hmm, just found out about kinda sad issue with this approach. When I want to set focus for something, I obviously would like to remove focus from remaining provider instances. Otherwise it would not be possible to refocus there again since state would still have a `true` there.

So is it possible to somehow find out all instances for dynamic providers so I can dispatch action to remove focus there?

**Comment by timbur** [comment]
I see.  My assumption was that you wanted to somehow maintain focus on multiple elements at a time.  Are you only focusing on things like inputs/selects/textareas?  If that's the case, then you'll only be able to focus on one at a time anyway, and would mean it makes the most sense to use a single static provider.  Dispatching an action to all provider instances of a certain type seems like an anti-pattern to me, as it indicates that you might as well just be using a single provider.

Before I can really help out, I need to know what exactly it is you want to do.

**Comment by FredyC** [comment]
Naming it focus wasn't probably best idea, think of it as "active". It's supposed to be used to tell component that user is working with it and it should behave differently. It's not just about form elements. Example is handling handle keyboard navigation input only when component is active.

I wanted to avoid single provider because of various reasons I've explained in a first comment here.

**Comment by timbur** [comment]
I would use a single provider but include a `merge` function instead of a `key`.  The `merge` property allows you to merge the provider's state with the component's props.  And this would give you full control over exactly which elements are active at any given time.  You would still be able to use the `focusable` prop in your components as you have it.

Something like this:
```js
// I have no idea how `redux-action-schema` works so I've removed it.
// Assuming `focusable` is your identifier for which elements are active,
// the idea here is that a `hasFocusMap` reducer is an object which maps
// the `focusable` identifiers to boolean values to indicate which elements are active.

const SET_FOCUS = 'SET_FOCUS';
const REMOVE_FOCUS = 'REMOVE_FOCUS';

const actions = {
  // `removeOthers` will be a boolean value that you can use to remove focus on all others
  setFocus: (focusable, removeOthers) => ({ type: SET_FOCUS, focusable, removeOthers }),
  removeFocus: (focusable, removeAll) => ({ type: REMOVE_FOCUS, focusable })
};

const reducers = {
  hasFocusMap(state = {}, action) {
    switch (action.type) {
      case SET_FOCUS:
        return action.removeOthers
          ? { [action.focusable]: true }
          : { ...state, [action.focusable]: true };

      case REMOVE_FOCUS:
        return action.removeAll
          ? {}
          : { ...state, [action.focusable]: false };

      default:
        return state;
    }
  }
};

const merge = {
  // This function will provide a single `hasFocus` boolean value
  // to your components based on a merging of their `focusable` prop
  // and the state of `hasFocusMap`.
  hasFocus: {
    // Need `keys` array for efficient updates (only runs `get` function when `hasFocusMap` changed).
    keys: ['hasFocusMap'],
    // The component(s) will also update only if the value returned by the `get` function has changed. (Efficient! 😉)
    get: (state, props) => Boolean(state.hasFocusMap[props.focusable])
  }
};

export default { actions, reducers, merge };
```

**Comment by FredyC** [comment]
Wow, that actually looks very useful. Thanks, I'll give it a spin.

**Comment by timbur** [comment]
Glad you finally find one of my suggestions useful!  Haha.  😀 😂 😂 😂 😁 

That is a really good idea about using `merge` to setup `setFocus` and `removeFocus`.  Let me know if you run into any problems with that.  And I'm curious to see how you might make it all work with `redux-action-schema`.  I'll definitely check that out.  👍

**Comment by timbur** [comment]
Make any progress on this?

**Comment by FredyC** [comment]
Haven't tried it yet actually, but I am getting there :)

**Comment by FredyC** [comment]
I think it's time to close this as I am not using providers anymore. Too complicated for my taste and my teammates were confused even more.

**Comment by timbur** [comment]
No problem.

I'm not sure how providers could be any simpler, as they are nothing more than states (as reducers) and actions to update those states. Anything more complicated than that usually means that you (not you specifically, but everyone using providers) should rethink their approach.

For what it's worth, I am planning on removing the built-in query handling and turning it into its own separate provider. Most people won't even need to use replication or many providers. Apologies for the confusion there.

I wish I had the time and energy to devote to further improve docs and examples to the same level of detail as vanilla redux. I will say though that I have been busy building things (real world applications) with providers and have not yet run into any limitations. Everything works perfectly on the first try and I highly doubt application features could be implemented faster and more reliably with any other paradigm (mobx included). It helps that I know how providers work in their entirety though of course.

I wish you the best of luck with vanilla redux or whatever flavor of flux you've gone with.

**Comment by FredyC** [comment]
I am not saying that providers are bad. It surely took a lot of work and thinking to put it together. But I see a major problem of adoption. Honestly, if someone asked me what are actual advantages of using providers over vanilla redux, I couldn't answer that well enough to convince them.

On the contrary, when I showed my project with providers to a coworker, he was totally confused on how it's possible that component is getting all those props, but there is no logic that would express that desire. As much as convenient this is once you know it, it's also very "magical" for a newcomer.
